README.md
# NativeScript Media Generator

[![Greenkeeper badge](https://badges.greenkeeper.io/hypery2k/nativescript-media-generator.svg)](https://greenkeeper.io/)
[![Build Status](https://travis-ci.org/hypery2k/nativescript-media-generator.svg?branch=master)](https://travis-ci.org/hypery2k/nativescript-media-generator) [![npm version](https://badge.fury.io/js/nativescript-media-generator.svg)](http://badge.fury.io/js/nativescript-media-generator) [![Dependency Status](https://david-dm.org/hypery2k/nativescript-media-generator.svg)](https://david-dm.org/hypery2k/nativescript-media-generator) [![devDependency Status](https://david-dm.org/hypery2k/nativescript-media-generator/dev-status.svg)](https://david-dm.org/hypery2k/nativescript-media-generator#info=devDependencies) 

[![Bountysource](https://www.bountysource.com/badge/tracker?tracker_id=11673029)](https://www.bountysource.com/trackers/11673029-hypery2k-nativescript-media-generatorerator?utm_source=11673029&utm_medium=shield&utm_campaign=TRACKER_BADGE) [![Flattr this git repo](http://api.flattr.com/button/flattr-badge-large.png)](https://flattr.com/submit/auto?user_id=mreinhardt&url=https://github.com/hypery2k/nativescript-media-generator&title=badges&language=&tags=github&category=software)

> CLI Utility that generates NativeScript image assets required for images, icons, and splash screens. 

[![NPM](https://nodei.co/npm/nativescript-media-generator.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/nativescript-media-generator/)

It requires your logo to have a solid background colour but does not distort or lose any image content so everything is at the maximum size without loss.

## Installation

Need imagemagick

On Debian/Ubuntu:

```bash
sudo apt-get install imagemagick
```

on OS X:

```bash
brew install imagemagick
```

```bash
$ npm install -g nativescript-media-generator
```

Once installed, cd to the root of your NativeScript application and run:
```bash
$ tns-media-gen <logofilename.jpg> <backgroundcolourinhex-egFFF>
```


Or if you want to use the development version (nightly build), which maybe not stable!:

```
tns plugin add nativescript-media-generator@next
```

### Usage

If using this plugin, we recommand an **1900x2850** Portrait Area as basis layout.

```bash
$ tns-media-gen logo.jpg fff
```

To create the config 

```bash
$ tns-media-gen init
```

If you have created a `mediagen-config.json` file (see below), you can just run:
```bash
$ tns-media-gen
```

This will overwrite all logos and splash screen images in the `<projectdir>/app` directory with the correct sizes and in the correct location for NativeScript (As at 3.5)
> The recommended image or logo size is 2000px x 2000px. Its not a problem if the logo isn't square.

It also creates a `<projectdir>/Media` directory that has images for the Apple and Android stores such as an app icon.

## Custom Assets
You can create additional custom images if you need to submit to alternative app stores or have other needs that we haven't thought of yet.

Simply go to your project directory and run
```bash
$ tns-media-gen init
```

It will create an example file called `mediagen-config.json` which you can now edit. Add as many or few files as you need to the array.

> Note: The default path for files is the `<projectdir>/platforms` directory, you might need to use `../` as in the example below

### Example `mediagen-config.json`
```javascript
{
  "mediaPath": "media",
  "icon": {
    "filename": "etc/images/icon.png",
    "background": "fff"
  },
  "splash": {
    "filename": "etc/images/splash.png",
    "background": "fff"
  },
  "images": [
    {
      "filename": "icon.png",
      "alias": "logo.png"
    },
    {
      "filename": "icon.png",
      "alias": "logo_login.png"
    }
  ]
}
```

## Screenshots (Experimental)
Screenshots are a new feature that use PhantomJS to quickly generate a number of the screenshot assets you need when submitting to the app store. You need your local development server running when you execute `mediagen` so it can access the NativeScript versions of the files. Of course this won't be able to access a number of things in lots of apps but hopefully it should get you up on the app store pretty quickly.
This is still pretty experimental, and also please be aware that it may seem to hang for a minute or so before you see the screenshots being generated (it does need to visit the page each time!)


## Config
The config variables are below:
- mediaPath: path to the output folder
- icon: icon image
    - filename: path to source filename
    - background: solid colour in hex
- splash: splash image
    - filename: path to source filename
    - background: solid colour in hex
- images: an array of images which should generate for both platforms (**NOTE**: The biggest resolutution (100%) needed is 640dpi)
    - filename: the output file name with extension
    - alias to use (target filename)
- custom images: an array of custom image objects for additional media if desired
    - width: the width of the image in pixels
    - height: the height of the image in pixels
    - path: the directory to save the output
    - filename: the output file name with extension
    - source: the source file, same definition as an icon or splash above
        - filename: path to source filename
        - background: solid colour in hex
- screenshots: An array of screenshot objects
    - url: url to your local development server
    - name: name of the page for easy reference later on


CHANGELOG.md
<a name="1.0.0"></a>
# [1.0.0](https://github.com/hypery2k/nativescript-media-generator/compare/v0.2.4...v1.0.0) (2018-01-08)


### Bug Fixes

* **config-error:** Resolve config not found error ([f2b6684](https://github.com/hypery2k/nativescript-media-generator/commit/f2b6684)), closes [#14](https://github.com/hypery2k/nativescript-media-generator/issues/14)
* **convert-error:** Resolve resize error without giving width ([e73bed6](https://github.com/hypery2k/nativescript-media-generator/commit/e73bed6)), closes [#14](https://github.com/hypery2k/nativescript-media-generator/issues/14)
* **node-env:** Correct command not found error ([4c264bc](https://github.com/hypery2k/nativescript-media-generator/commit/4c264bc)), closes [#14](https://github.com/hypery2k/nativescript-media-generator/issues/14)
* **phantom-usage:** Use phantom 3.x to support node4+ ([2af5e89](https://github.com/hypery2k/nativescript-media-generator/commit/2af5e89))


### Features

* **iPhoneX:** Support for iPhone X ([2c46388](https://github.com/hypery2k/nativescript-media-generator/commit/2c46388))



<a name="0.2.4"></a>
## [0.2.4](https://github.com/hypery2k/nativescript-media-generator/compare/v0.2.3...v0.2.4) (2017-03-02)


### Bug Fixes

* **build:** Fix build error ([4ce81ae](https://github.com/hypery2k/nativescript-media-generator/commit/4ce81ae))
* **build:** Fix build error ([cbf9d82](https://github.com/hypery2k/nativescript-media-generator/commit/cbf9d82))
* **cli:** Fix binary command error ([a54e067](https://github.com/hypery2k/nativescript-media-generator/commit/a54e067)), closes [#13](https://github.com/hypery2k/nativescript-media-generator/issues/13)



<a name="0.2.3"></a>
## [0.2.3](https://github.com/hypery2k/nativescript-media-generator/compare/v0.2.2...v0.2.3) (2017-01-30)


### Bug Fixes

* **node4:** Use older node for node4+ support ([a8ce11c](https://github.com/hypery2k/nativescript-media-generator/commit/a8ce11c))
* **resizing:** Correct percentages for images ([27f7ad3](https://github.com/hypery2k/nativescript-media-generator/commit/27f7ad3))



<a name="0.2.2"></a>
## [0.2.2](https://github.com/hypery2k/nativescript-media-generator/compare/v0.2.1...v0.2.2) (2017-01-11)


### Bug Fixes

* **splash-screen:** Correct android splash screen size ([f4f2d71](https://github.com/hypery2k/nativescript-media-generator/commit/f4f2d71))



<a name="0.2.1"></a>
## [0.2.1](https://github.com/hypery2k/nativescript-media-generator/compare/6aedc62...v0.2.1) (2016-12-13)


### Bug Fixes

* **image-gen:** Correct property usage for image generation ([b21d7ce](https://github.com/hypery2k/nativescript-media-generator/commit/b21d7ce))
* **module_missing:** Add missing async module ([6aedc62](https://github.com/hypery2k/nativescript-media-generator/commit/6aedc62))
* **test:** Fix test errors ([a53c3b2](https://github.com/hypery2k/nativescript-media-generator/commit/a53c3b2))
* **test:** FIx test errors ([3223ff6](https://github.com/hypery2k/nativescript-media-generator/commit/3223ff6))


### Features

* **additional_images:** Allow additional images ([39673ca](https://github.com/hypery2k/nativescript-media-generator/commit/39673ca))
* **images:** Support for images ([b2ba106](https://github.com/hypery2k/nativescript-media-generator/commit/b2ba106))





test/specs/app.spec.js
var mediaGen = require('../../tns-media-gen');

describe('main app', function() {
    describe('resize', function() {
        it('should not error out', function(done) {
            mediaGen.__resize(10, 10, "#fff", "images/icon.png", "output.png", "../../.tmp")
                .then(function(result) {
                    expect(result).toBeDefined();
                    done();
                })
                .catch(done);
        })
    });

    describe('generate', function() {
        it('should not error out', function(done) {
            mediaGen.__generate()
                .then(function(result) {
                    done();
                })
                .catch(done)
        })
    });

    describe('genConfig', function() {
        it('should not error out', function(done) {
            mediaGen.__genConfig()
                .then(function(result) {
                    expect(result).toBe('success');
                    done();
                })
                .catch(done);
        })
    })
});

test/specs/screenshots.spec.js
var screenshots = require('../../screenshots'),
  fs = require('fs');

var settings = {
  url: 'http://www.google.com',
  width: 1280,
  height: 720,
  path: "images/Google",
  fileName: '7Inch.png'
};

xdescribe('Screenshots', function () {
  describe('generate', function () {
    it('should return no errors', function (done) {
      screenshots.generate(settings.url, settings.width, settings.height, 1, settings.path, settings.fileName)
        .then(function (result) {
          console.log('==>' + result)
          expect(result.success).toBeDefined();
        })
        .catch(function (error) {
          throw new error;
        })
        .done(function () { done() });
    }, 30000);

    it('should have created a file', function (done) {
      fs.exists(settings.path + "/" + settings.fileName, function (result) {
        expect(result).toBeTruthy();
        done();
      });
    }, 30000);

    it('should die when nothing is supplied', function (done) {
      screenshots.generate()
        .then(function (result) {
          fail("Should not have worked");
        })
        .catch(function (err) {
          expect(err).toBeDefined();
          done();
        });
    }, 30000);
  });

  describe('generate all', function () {
    screenshots.pages.push({ url: "http://www.google.com", name: "home" });

    it('should return', function (done) {
      expect(screenshots.generateAll()).toBeTruthy();
      done();
    }, 30000);

    it('should have created a file', function (done) {
      fs.exists("Media/android/screenshots/10in/android-10in-1280x720-google.png", function (result) {
        expect(result).toBeTruthy();
        done();
      });
    }, 30000);
  });

});


lib/resize.js
 var q = require('q'),
     path = require('path'),
     fs = require('fs'),
     gm = require('gm').subClass({ imageMagick: true }),
     mkdirp = require('mkdirp');

 module.exports = function resize(width, height, bgColour, imagePath, outputFilename, outputPath) {
     var deferred = q.defer(),
         filepath = path.join(process.cwd(), imagePath);
     gm(filepath).size(function(error, size) {

         if (error) {
             console.error("GM Error", error);
             deferred.reject(error);
         } else {

             // current image size
             var imageWidth = size.width;
             var imageHeight = size.height;
             var ratio = width / height,
                 imageRatio = imageWidth / imageHeight;
             var image = this;

             if (isNaN(parseFloat(width))) {
                 var percentage = '' + width;
                 width = imageWidth * (parseInt(percentage.substring(0, percentage.indexOf("%")), 0) / 100);
             }
             // center placement
             if (ratio >= 1) {
                 //Landscape or square
                 var newWidth = height * imageRatio;

                 if (newWidth >= width) {
                     this.resize(width);
                 } else {
                     this.resize(null, height);
                 }


             } else {
                 var newHeight = width / imageRatio;

                 if (newHeight >= height) {
                     this.resize(null, height);
                 } else {
                     this.resize(width);
                 }

             }

             var x = (width / 2) - (imageWidth / 2);
             var y = (height / 2) - (imageHeight / 2);

             mkdirp(path.join(process.cwd(), "app", "App_Resources", outputPath), function() {
                 image.background(bgColour)
                     .gravity('Center')
                     .extent(width, height)
                     .write(path.join(process.cwd(), "app", "App_Resources", outputPath, outputFilename), function(error) {
                         if (error) {
                             console.error("Write file error", error);
                             deferred.reject(error);
                         } else {
                             console.log(this.outname);
                             deferred.resolve(this.outname);
                         }
                     });
             });
         }

     });
     return deferred.promise;
 }

screenshots.js
'use strict';

var phantom = require('phantom'),
  path = require('path'),
  q = require('q'),
  mkdirp = require('mkdirp'),
  async = require('async'),
  pkg = require('./package.json'),
  config = require(process.cwd() + "/mediagen-config");

var ph;
var mediaPath = config.mediaPath || 'Media';

function evaluate(page, func) {
  var args = [].slice.call(arguments, 2);
  var fn = "function() { return (" + func.toString() + ").apply(this, " + JSON.stringify(args) + ");}";
  return page.evaluate(fn);
}

var output = {
  generate: function (url, width, height, devicePixelRatio, savePath, saveFilename) {
    var deferred = q.defer();

    if (!url || !width || !height || !savePath || !saveFilename) {
      deferred.reject({
        error: "A required argument is missing"
      });
      return deferred.promise;
    }
    if (!ph) {
      phantom.create(function (newPh) {
        ph = newPh;
        doGen();
      })
    } else {
      doGen();
    }

    function doGen() {
      ph.createPage(function (page) {
        page.setViewportSize(width, height, function () {
          page.open(url, function () {
            if (devicePixelRatio && devicePixelRatio !== 1) {
              evaluate(page, function (devicePixelRatio) {
                document.body.style.webkitTransform = "scale(" + devicePixelRatio + ")";
                document.body.style.webkitTransformOrigin = "0% 0%";
                document.body.style.width = (100 / devicePixelRatio) + "%";
              }, devicePixelRatio);
            }
            mkdirp(path.resolve(__dirname, savePath), function (err) {
              if (err) {
                deferred.reject(err);
              }
              page.render(savePath + "/" + saveFilename, {
                format: 'jpg',
                quality: '60'
              }, function (err) {
                console.log("Generated screenshot", url, savePath, saveFilename);
                if (err) {
                  deferred.reject(err);
                } else {
                  deferred.resolve({
                    success: true
                  });
                }
                page.close();
              });
            });
          });
        });
      });
    }

    return deferred.promise;
  },
  generateAll: function () {
    var deferred = q.defer();
    async.eachLimit(output.screenshots, 1, function (item, cb1) {
      async.eachLimit(output.pages, 1, function (page, cb2) {
        output.generate(page.url, item.width, item.height, item.devicePixelRatio, item.savePath, item.filename + page.name + ".jpg")
          .then(function (result) {
            cb2();
          })
          .catch(cb2)
      }, function (err) {
        cb1(err);
      })
    }, function (err) {
      if (ph) ph.exit();
      if (err) {
        deferred.reject(err);
      } else {
        deferred.resolve(true);
      }
    });
    return deferred.promise;
  },
  pages: [],
  screenshots: [{
    filename: "android-10in-1280x720-land",
    width: 1280,
    height: 720,
    savePath: mediaPath + '/android/screenshots/10in'
  }, {
    filename: "android-10in-2048x1152-land",
    width: 2048,
    height: 1152,
    savePath: mediaPath + '/android/screenshots/10in'
  }, {
    filename: "android-7in-1280x800-land",
    width: 1280,
    height: 800,
    savePath: mediaPath + '/android/screenshots/7in'
  }, {
    filename: "android-10in-1280x720-port",
    width: 720,
    height: 1280,
    savePath: mediaPath + '/android/screenshots/10in'
  }, {
    filename: "android-10in-2048x1152-port",
    width: 1152,
    height: 2048,
    savePath: mediaPath + '/android/screenshots/10in'
  }, {
    filename: "android-7in-1280x800-port",
    width: 800,
    height: 1280,
    savePath: mediaPath + '/android/screenshots/7in'
  }, {
    filename: "android-4in-1280x720-land",
    width: 1920,
    height: 1080,
    savePath: mediaPath + '/android/screenshots/4in'
  }, {
    filename: "android-4in-1280x720-port",
    width: 1080,
    height: 1920,
    savePath: mediaPath + '/android/screenshots/4in'
  }, {
    filename: "ipad-1024x768-land",
    width: 1024,
    height: 768,
    savePath: mediaPath + '/ios/screenshots/ipad'
  }, {
    filename: "ipadretina-2048x1536-land",
    width: 2048,
    height: 1536,
    devicePixelRatio: 2,
    savePath: mediaPath + '/ios/screenshots/ipadRetina'
  }, {
    filename: "iphone4-960x640-land",
    width: 960,
    height: 640,
    savePath: mediaPath + '/ios/screenshots/iphone4'
  }, {
    filename: "iphone5-1136x640-land",
    width: 1136,
    height: 640,
    savePath: mediaPath + '/ios/screenshots/iphone5'
  }, {
    filename: "ipad-1024x768-port",
    width: 768,
    height: 1024,
    savePath: mediaPath + '/ios/screenshots/ipad'
  }, {
    filename: "ipadretina-2048x1536-port",
    width: 1536,
    height: 2048,
    devicePixelRatio: 2,
    savePath: mediaPath + '/ios/screenshots/ipadRetina'
  }, {
    filename: "iphone4-640x960-port",
    width: 640,
    height: 960,
    savePath: mediaPath + '/ios/screenshots/iphone4'
  }, {
    filename: "iphone5-640x1136-port",
    width: 640,
    height: 1136,
    savePath: mediaPath + '/ios/screenshots/iphone5'
  }, {
    filename: "iphone6p-1242x2208-port",
    width: 1242,
    height: 2208,
    savePath: mediaPath + '/ios/screenshots/iphone6p'
  }, {
    filename: "iphone6p-2208x1242-land",
    width: 2208,
    height: 1242,
    savePath: mediaPath + '/ios/screenshots/iphone6p'
  }, {
    filename: "iphone6-1334x750-land",
    width: 1334,
    height: 750,
    savePath: mediaPath + '/ios/screenshots/iphone6'
  }, {
    filename: "iphone6-750x1334-port",
    width: 750,
    height: 1334,
    savePath: mediaPath + '/ios/screenshots/iphone6'
  }]
};
module.exports = output;


lib/store_images.js
 module.exports = function storeImagesGenerator(config, mediaPath) {
     return [
         //Android Store Icons
         {
             width: 512,
             height: 512,
             path: "../../" + mediaPath + "/android/store",
             filename: "512.png",
             source: process.argv[2] || config.icon || config.image
         }, {
             width: 1024,
             height: 500,
             path: "../../" + mediaPath + "/android/store",
             filename: "1024x500.png",
             source: process.argv[2] || config.splash || config.image
         }, {
             width: 180,
             height: 120,
             path: "../../" + mediaPath + "/android/store",
             filename: "180x120.png",
             source: process.argv[2] || config.splash || config.image
         },

         //Apple store icons
         {
             width: 1024,
             height: 1024,
             path: "../../" + mediaPath + "/ios/store",
             filename: "1024x1024-AppIcon.jpg",
             source: process.argv[2] || config.icon || config.image
         }
     ];
 }

lib/splash_images.js
module.exports = function splashImagesGenerator(config, mediaPath) {
    return [
        //iOS Splash
        {
            width: 640,
            height: 1136,
            path: "iOS/Assets.xcassets/LaunchImage.launchimage",
            filename: "Default-568h@2x.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 1024,
            height: 768,
            path: "iOS/Assets.xcassets/LaunchImage.launchimage",
            filename: "Default-Landscape.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 1536,
            height: 2048,
            path: "iOS/Assets.xcassets/LaunchImage.launchimage",
            filename: "Default-Portrait@2x.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 768,
            height: 1024,
            path: "iOS/Assets.xcassets/LaunchImage.launchimage",
            filename: "Default-Portrait.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 640,
            height: 960,
            path: "iOS/Assets.xcassets/LaunchImage.launchimage",
            filename: "Default@2x.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 320,
            height: 480,
            path: "iOS/Assets.xcassets/LaunchImage.launchimage",
            filename: "Default.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 750,
            height: 1344,
            path: "iOS/Assets.xcassets/LaunchImage.launchimage",
            filename: "Default-667h.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 1242,
            height: 2208,
            path: "iOS/Assets.xcassets/LaunchImage.launchimage",
            filename: "Default-736h.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 2208,
            height: 1242,
            path: "iOS/Assets.xcassets/LaunchImage.launchimage",
            filename: "Default-Landscape-736h.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 768,
            height: 1024,
            path: "iOS/Assets.xcassets/LaunchScreen.AspectFill.imageset",
            filename: "LaunchScreen-AspectFill.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 1536,
            height: 2048,
            path: "iOS/Assets.xcassets/LaunchScreen.AspectFill.imageset",
            filename: "LaunchScreen-AspectFill@2x.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 384,
            height: 512,
            path: "iOS/Assets.xcassets/LaunchScreen.Center.imageset",
            filename: "LaunchScreen-Center.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 768,
            height: 1024,
            path: "iOS/Assets.xcassets/LaunchScreen.Center.imageset",
            filename: "LaunchScreen-Center@2x.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 1125,
            height: 2436,
            path: "iOS/Assets.xcassets/LaunchImage.launchimage",
            filename: "Default-1125h.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 2436,
            height: 1125,
            path: "iOS/Assets.xcassets/LaunchImage.launchimage",
            filename: "Default-Landscape-X.png",
            source: process.argv[2] || config.splash || config.image
        },
        // Android
        {
            width: 480,
            path: "Android/drawable-mdpi",
            filename: "splash.png",
            source: process.argv[2] || config.splash || config.image
        },
        {
            width: 800,
            path: "Android/drawable-hdpi",
            filename: "splash.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 960,
            path: "Android/drawable-xhdpi",
            filename: "splash.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 1600,
            path: "Android/drawable-xxhdpi",
            filename: "splash.png",
            source: process.argv[2] || config.splash || config.image
        }, {
            width: 1920,
            path: "Android/drawable-xxxhdpi",
            filename: "splash.png",
            source: process.argv[2] || config.splash || config.image
        }

    ];
}

lib/generate_image_sets.js
 var q = require('q'),
     path = require('path'),
     fs = require('fs'),
     gm = require('gm').subClass({ imageMagick: true });

 module.exports = function generateImageSets(filename, source, iOSPath) { /*eslint complexity: [error, 22]*/
     return [
         //IOS Images
         {
             width: "24.5%", // max 466
             path: 'iOS/Assets.xcassets' + (iOSPath ? '/' + iOSPath : ''),
             filename: filename,
             source: source
         }, {
             width: "49%", // max 932
             path: 'iOS/Assets.xcassets' + (iOSPath ? '/' + iOSPath : ''),
             filename: filename.replace('.', '@2x.'),
             source: source
         }, {
             width: "73.6%", // max 1398
             path: 'iOS/Assets.xcassets' + (iOSPath ? '/' + iOSPath : ''),
             filename: filename.replace('.', '@3x.'),
             source: source
         },

         //Android
         {
             width: "75%", // max 1425
             path: "Android/drawable-hdpi",
             filename: filename,
             source: source
         }, {
             width: "50%", // max 950
             path: "Android/drawable-mdpi",
             filename: filename,
             source: source
         }, {
             width: "100%", // max 1900
             path: "Android/drawable-xhdpi",
             filename: filename,
             source: source
         }, {
             width: "70%", // max 1331
             path: "Android/drawable-xxhdpi",
             filename: filename,
             source: source
         }, {
             width: "93.4%", // max 1775
             path: "Android/drawable-xxxhdpi",
             filename: filename,
             source: source
         }

     ]
 }

tns-media-gen.js
#!/usr/bin/env node

"use strict";

var gm = require('gm').subClass({ imageMagick: true }),
    mkdirp = require('mkdirp'),
    path = require('path'),
    fs = require('fs'),
    screenshots = require('./screenshots'),
    q = require('q'),
    resize = require('./lib/resize'),
    generateImageSets = require('./lib/generate_image_sets'),
    generateIconSets = require('./lib/generate_icon_sets'),
    storeImagesGenerator = require('./lib/store_images'),
    splashImagesGenerator = require('./lib/splash_images'),
    config,
    mediaPath;


function generate() {
    var deferred = q.defer();

    fs.readdir(path.join(process.cwd(), "app", "App_Resources", "iOS"), function(err, result) { /*eslint complexity: [error, 55]*/

        if (err) {
            throw err;
        } else {

            if (!config && config.image) config.image = {
                filename: "./icon.png"
            };

            var storeImages = storeImagesGenerator(config, mediaPath),
                splashImages = splashImagesGenerator(config, mediaPath),
                source = process.argv[2] || config.icon || config.image,
                images = generateIconSets('icon-20.png', source, 'AppIcon.appiconset')

            images.push({
                width: 1024,
                height: 1024,
                path: 'iOS/Assets.xcassets/AppIcon.appiconset',
                filename: "icon-1024.png",
                source: source
            });
            images.push({
                width: 180,
                height: 180,
                path: 'iOS/Assets.xcassets/AppIcon.appiconset',
                filename: "icon-60@3x.png",
                source: source
            });

            images = images.concat(splashImages);

            if (config.screenshots && config.screenshots.length) {
                screenshots.pages = config.screenshots;
            }


            if (config.images) {
                config.images.forEach(function(item) {
                    var additionalImages = generateImageSets(item.alias ? item.alias : item.filename, { filename: item.filename });
                    additionalImages.forEach(function(additionalImage) {
                        images.push(additionalImage);
                    });
                });
            }

            if (config.customImages) {
                config.customImages.forEach(function(item) {
                    images.push(item);
                });
            }


            if (!process.argv[2] && !config.image && (!config.icon && !config.splash)) {
                console.log('You must specify a filename as the second argument, or in a config file');
            } else if (!process.argv[3] && !config.background && (!config.icon && !config.splash)) {
                console.log("Please specify a background colour in hex values as the third argument, or in a config file");
            } else {

                var totalImages = images.length + (screenshots.pages.length * screenshots.screenshots.length);
                console.log("------------------------------");
                console.log("   nativescript-media-generator");
                console.log("------------------------------");
                console.log("Generating " + totalImages + " images so you don't have to");
                console.log("------------------------------");
                screenshots.generateAll();
                images.forEach(function(image) {
                    var background, sourceImage;
                    if (process.argv[2]) {
                        /*
                         Legacy Configuration
                         */
                        if (process.argv[3]) {
                            background = process.argv[3];
                        } else {
                            background = config.background;
                        }
                        if (process.argv[2]) {
                            sourceImage = process.argv[2];
                        } else {
                            sourceImage = config.image;
                        }
                    } else {
                        if (image.source) {
                            sourceImage = image.source.filename;
                            background = image.source.background;
                        }
                    }
                    if (sourceImage) {
                        resize(image.width, image.height, '#' + background, sourceImage, image.filename, image.path);

                    }
                });
                deferred.resolve();
            }
        }
    });
    return deferred.promise;
}

function genConfig() {
    var deferred = q.defer();
    var destFile = path.join(process.cwd(), "mediagen-config.json");

    fs.writeFile(destFile, JSON.stringify({
        "icon": { "filename": "icon.png", "background": "fff" },
        "splash": { "filename": "splash.png", "background": "fff" },
        "customImages": [{
            "width": 120,
            "height": 120,
            "path": "../../" + mediaPath + "/custom",
            "filename": "outputFilename.png",
            "source": { "filename": "image.png", "background": "fff" }
        }],
        "screenshots": [{ "url": "http://www.google.com", "name": "homepage" }]
    }, null, 4), function() {
        deferred.resolve("success");
    });
    console.log("Created `mediagen-config.json` file in the current directory.");
    return deferred.promise;
}

try { /*eslint global-require: off*/
    config = require(process.cwd() + "/mediagen-config");
    mediaPath = config.mediaPath || 'Media';
} catch (e) {
    if (process.argv[2] !== "init") {
        console.log("Could not find configuration file. To create one run `$ mediagen init`");
    }
}

switch (process.argv[2]) {
    case "init":
        genConfig();
        break;
    default:
        generate();
}

module.exports = {
    __resize: resize,
    __generate: generate,
    __genConfig: genConfig
};